src/metrics.cpp

```diff
@@ -0,0 +1,75 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#include <PowerWAF.hpp>
+#include <metrics.hpp>
+
+namespace ddwaf
+{
+
+ddwaf_metrics metrics_collector::generate_metrics()
+{
+    ddwaf_metrics final_metrics = { total_runtime_, { 0 } };
+    ddwaf_object_map(&final_metrics.rule_runtime);
+
+    for (unsigned i = 0; i < rule_runtime_.size(); i++)
+    {
+        if (rule_runtime_[i] == 0)
+        {
+            continue;
+        }
+
+        // TODO this is weird
+        const std::string& rule_id = rules_[i].id;
+        ddwaf_object duration;
+        ddwaf_object_unsigned_force(&duration, rule_runtime_[i]);
+        ddwaf_object_map_addl(&final_metrics.rule_runtime,
+                              rule_id.c_str(), rule_id.size(), &duration);
```

Comment:
I don't like that this will do allocations and copy every single rule name. We're talking 100+ allocations+copy that will have to be done on every request. Also the final size of the map is known in advance, but the ddwaf_objects are in the map are not preallocated.

Reply:
Actually the size of the map is not known beforehand as we only include those with non-zero duration.